ddtrace/contrib/django/patch.py

```diff
@@ -666,14 +674,187 @@ def django_asgi_modifier(span, scope):
+        return getattr(user, user_type.EMAIL_FIELD, None)
+
+    return _find_in_user_model(user, _POSSIBLE_EMAIL_FIELDS)
+
+
+def _get_name(user):
+    if hasattr(user, "get_full_name"):
+        return user.get_full_name()
+
+    if hasattr(user, "first_name") and hasattr(user, "last_name"):
+        return "%s %s" % (user.first_name, user.last_name)
+
+    return getattr(user, "name", None)
+
+
+def _get_user_info(user):
+    """
+    In safe mode, try to get the user id from the user object.
+    In extended mode, try to also get the username (which will be the returned user_id),
+    email and name.
+
+    Since the Django Authentication model supports pluggable users models
+    there is some heuristic involved in extracting this field, though it should
+    work well for the most common case with the default User model.
+    """
+    user_extra_info = {}
+
+    if config._automatic_login_events_mode == "safe":
+        user_id = _get_userid(user)
+    else:  # extended mode, default
+        user_id = _get_username(user)
+        if not user_id:
+            user_id = _find_in_user_model(user, _POSSIBLE_USER_ID_FIELDS)
+
+        user_extra_info = {
+            "login": user_id,
+            "email": _get_user_email(user),
+            "name": _get_name(user),
+        }
+
+    if not user_id:
+        return None, {}
+
+    return user_id, user_extra_info
+
+
+@trace_utils.with_traced_module
+def traced_login(django, pin, func, instance, args, kwargs):
+    func(*args, **kwargs)
+
+    try:
+        mode = config._automatic_login_events_mode
+        request = get_argument_value(args, kwargs, 0, "request")
+        user = get_argument_value(args, kwargs, 1, "user")
+
+        if not config._appsec_enabled or mode == "disabled":
+            return
+
+        if user and str(user) != "AnonymousUser":
+            user_id, user_extra = _get_user_info(user)
+            if not user_id:
+                log.debug(
+                    "Automatic Login Events Tracking: " "Could not determine user id field user for the %s user Model",
+                    type(user),
+                )
+                return
+
+            with pin.tracer.trace("django.contrib.auth.login", span_type=SpanTypes.AUTH):
+                from ddtrace.contrib.django.compat import user_is_authenticated
+
+                if user_is_authenticated(user):
+                    session_key = getattr(request, "session_key", None)
+                    track_user_login_success_event(
+                        pin.tracer,
+                        user_id=user_id,
+                        session_id=session_key,
+                        propagate=True,
+                        login_events_mode=mode,
+                        **user_extra
+                    )
+                    return
+
+                # Login failed but the user exists
+                track_user_login_failure_event(pin.tracer, user_id=user_id, exists=True, login_events_mode=mode)
```

Comment:
Would moving the above return to below `track_user_login_failure_event()`  be helpful here? I just worry about modifying this function in the future and getting confusing behavior if you added code outside of the else statement since it would run, when we would have wanted to return already.
```suggestion
                else:
                        track_user_login_failure_event(pin.tracer, user_id=user_id, exists=True, login_events_mode=mode)
                return
```

Reply:
Yes, being explicit is probably better in this case, commited plus another explicit return on the other failure track call below that one.